File: fastpowerset.go
func subsets(set []int) [][]int {
    n := len(set)
    result := make([][]int, 0, 1<<uint(n))

    for i := 0; i < (1 << uint(n)); i++ {
        subset := make([]int, 0)
        for j := 0; j < n; j++ {
            if i&(1<<uint(j)) != 0 {
                subset = append(subset, set[j])
            }
        }
        result = append(result, subset)
    }

    return result
}

File: fastpowerset.rs
fn power_set(set: &[i32]) -> Vec<Vec<i32>> {
    let n = set.len();
    let mut result = Vec::with_capacity(1 << n);

    for i in 0..(1 << n) {
        let mut subset = Vec::new();
        for j in 0..n {
            if (i & (1 << j)) != 0 {
                subset.push(set[j]);
            }
        }
        result.push(subset);
    }

    result
}

fn main() {
    let set = vec![1, 2, 3];
    let result = power_set(&set);
    println!("{:?}", result);
    // Output: [[], [1], [2], [1, 2], [3], [1, 3], [2, 3], [1, 2, 3]]
}
